src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/PropagateCancellationTokensWhenPossible.cs

```diff
@@ -33,11 +39,108 @@ public abstract class PropagateCancellationTokensWhenPossibleAnalyzer : Diagnost
 
         public override void Initialize(AnalysisContext analysisContext)
         {
-            // TODO: Enable concurrent execution of analyzer actions.
-            //analysisContext.EnableConcurrentExecution();
+            analysisContext.EnableConcurrentExecution();
 
-            // TODO: Configure generated code analysis.
-            //analysisContext.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            // TODO: Include this?
+            // analysisContext.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            analysisContext.RegisterOperationAction(AnalyzeInvocation, OperationKind.Invocation);
+        }
+
+        private static void AnalyzeInvocation(OperationAnalysisContext context)
+        {
+            var invocation = (IInvocationOperation)context.Operation;
+            if (InvocationShouldPassCancellationToken(invocation, context.Compilation))
+            {
+                context.ReportDiagnostic(invocation.Syntax.CreateDiagnostic(Rule));
```

Comment:
We should update the message to tell the user what cancellation token they should be passing (ie the field or parameter).

Reply:
I was planning to offer multiple codefixes, one for each cancellation token in scope; that's why I avoided mentioning a specific one in the analyzer message.

Reply:
@333fred

> and we can put all the valid ones in the message.

That might be messy because there isn't a neat way to pass a variadic number of format arguments to the message. e.g.

    // Can't design a single message format to cover all of these cases
    // ("Propagate CancellationToken {0}" might work, but that would interfere with localization because
    // the program would have to generate 'or' in English)
    Propagate CancellationToken 'ct1'
    Propagate CancellationToken 'ct1' or 'ct2'
    Propagate CancellationToken 'ct1', 'ct2', or 'ct3'

How about I show the name only when there's one `CancellationToken` in scope?

Reply:
There's nothing saying we have to use or, we should be able to do something like `Propagate CancellationTokens. Tokens in scope are: ct1, ct2, ct3`

Reply:
Ok. I'll settle on `Propagate CancellationToken in method invocation. Available tokens: 'ct1', 'ct2', 'ct3'`